### DISCUSSION TITLE
What is the proper way of defining visibility/import/export macros in meson?

### DISCUSSION THREAD

**Discussion by MehradN** [mesonbuild/meson#13626] [Q&A]
I'm developing a cross-platform library for Linux and Windows, which should be built and linked as a shared library. (But I want to allow building and linking as static library as well.)

I'm trying to follow [this guide](https://gcc.gnu.org/wiki/Visibility) (more specifically the "Step-by-step guide" section). But it requires a macro like `FOX_DLL` that is defined when the library is built and linked as a shared library, and a macro like `FOX_DLL_EXPORTS` that is defined when the library is only being built as a shared library.

I have some ideas on how to achieve this with Meson, but I'm not sure if I'm doing it correctly or if there is a better way to do it.

I know that the solution most likely revolves around `library() <lang>_shared_args`, `declare_dependency() compile_args/link_args`, `add_project_arguments` and `get_option('default_library')`. But I'm really confused, and I'm not sure which macros should be defined with which methods.

How should I define those macros? Should I follow other solutions for defining visibility macros, instead of the one mentioned in that guide?

**Comment by bruchar1** [comment]
With that example, you need to define `FOX_DLL` if you compile a shared library, and `FOX_DLL_EXPORTS` when you compile the actual DLL.

I will suppose `default_library` is either `static` or `shared`, as `both` would add more complexity... And I suppose language is cpp (but it could be c as well)

```
if get_option('default_library') == 'shared'
    add_project_arguments('-DFOX_DLL', language: 'cpp')
endif

foxlib = library(
    ...
    cpp_args: ['-DFOX_DLL_EXPORTS']  # not needed for static, but not harmful in that context
)

foxlib_dep = declare_dependency(
    link_with: foxlib,
)
```

**Reply by dcbaker** [reply]
Alternatively, if you're willing to use a more recent Meson (>= 1.3 I think), you can do:
```meson
foxlib = library(
  ...
   cpp_shared_args : ['-DFOX_DLL', '-DFOX_DLL_EXPORTS'],
)
```

Which will work correctly for shared, static, and both libraries

**Reply by bruchar1** [reply]
But you will still need to have `FOX_DLL` defined when you use the lib as a shared lib.

**Reply by dcbaker** [reply]
ah, that's true. This really gets back to the whole "we need to have shared vs static libs in `declare_dependency` so you could just put it there. So many interesting ways that `both` libs don't work right yet :/

**Reply by MehradN** [reply]
So when I add an argument with `add_project_arguments`, all projects that will use my library as their subproject dependency will be forced to use those arguments as well?

**Reply by MehradN** [reply]
Can the `both` library problem be fixed with checking `get_option('default_library') == 'both' and not get_option('prefer_static')`?

**Reply by MehradN** [reply]
> So when I add an argument with `add_project_arguments`, all projects that will use my library as their subproject dependency will be forced to use those arguments as well?

I tested it, and `add_project_arguments` doesn't force the dependents to define the macro.

**Reply by bruchar1** [reply]
`add_project_arguments` will add the args to all targets in your project. For targets outside your project, you are better to use the `compile_args` argument of `declare_dependency`
